### PR TITLE
Add paper's pre spawn events

### PR DESCRIPTION
**Pull request by Fusezion**
Stupid system forcing a creation

This PRs adds paper's three events for pre spawning 
- PreCreatureSpawnEvent (called before bukkit's creature spawn event)
- PreSpawnerSpawnEvent (called before bukkit's spawner spawn event)
- PhantomPreSpawnEvent (unsure much use for it but as it existed added it)

In addition added an effect for aborting the spawn events, if used the entity will not attempt to respawn again.

Didn't want to add this with the spawner support PR so waited until that was merged

### CODE REVIEW

**Comment by ShaneBeee** [comment]
Im genuinely curious, what is the point of these events?
Like what purpose do they server?

File: src/main/java/com/shanebeestudios/skbee/elements/other/effects/EffAbortSpawn.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+import ch.njol.skript.lang.Expression;
+import ch.njol.skript.lang.SkriptParser.ParseResult;
+import ch.njol.util.Kleenean;
+import com.destroystokyo.paper.event.entity.PreCreatureSpawnEvent;
+import org.bukkit.event.Event;
+import org.jetbrains.annotations.Nullable;
+
+@Name("Abort Creature Spawn")
+@Description({"Abort a creature spawn event, preventing it from retrying more attempts after canceling.",
+        "aborting should only be done if you're blanket blocking this entity type from spawning."})
```
Comment (ShaneBeee) [inline src/main/java/com/shanebeestudios/skbee/elements/other/effects/EffAbortSpawn.java:18]:
```suggestion
        "Aborting should only be done if you're blanket blocking this entity type from spawning."})
```

File: src/main/java/com/shanebeestudios/skbee/elements/other/effects/EffAbortSpawn.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+@Examples({"on pre creature spawn of an animal",
```
Comment (ShaneBeee) [inline src/main/java/com/shanebeestudios/skbee/elements/other/effects/EffAbortSpawn.java:19]:
```suggestion
@Examples({"on pre creature spawn of an animal:",
```

File: src/main/java/com/shanebeestudios/skbee/elements/other/effects/EffAbortSpawn.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        "\tcancel event",
+        "\tabort creature spawn"})
+@Since("INSERT VERSION")
+public class EffAbortSpawn extends Effect {
+
+    private static final boolean HAS_PRE_SPAWN_EVENT = Skript.classExists("com.destroystokyo.paper.event.entity.PreCreatureSpawnEvent");
+
+    static {
+        Skript.registerEffect(EffAbortSpawn.class,
+                "[:un]abort creature spawn");
```
Comment (ShaneBeee) [inline src/main/java/com/shanebeestudios/skbee/elements/other/effects/EffAbortSpawn.java:29]:
```suggestion
        Skript.registerEffect(EffAbortSpawn.class, "[:un]abort creature spawn");
```

File: src/main/java/com/shanebeestudios/skbee/elements/other/effects/EffAbortSpawn.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    protected void execute(Event event) {
+        if (event instanceof PreCreatureSpawnEvent preCreatureSpawnEvent)
+            preCreatureSpawnEvent.setShouldAbortSpawn(abort);
+    }
+
+    @Override
+    public String toString(@Nullable Event event, boolean b) {
+        return (abort ? "" : "un") + "abort creature spawn";
+    }
+}
```
Comment (ShaneBeee) [inline src/main/java/com/shanebeestudios/skbee/elements/other/effects/EffAbortSpawn.java:58]:
```suggestion
    }

}
```

**Comment by Fusezion** [comment]
> Im genuinely curious, what is the point of these events? Like what purpose do they server?

The pre spawner was one I need so didn't just want to only add it say you want to make a spawner plugin that not only stacks but also holds all drops inside of the spawner's nbt well you don't want to use `on spawner spawn` as the entity exists at the time causing you to worry about too many spawning in at once. So instead you'll want to use the loottable systems and cancel the spawning in the pre spawn this way the entity doesn't exist and no entity server lag

File: src/main/java/com/shanebeestudios/skbee/elements/other/events/EvtPreSpawn.java [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+import org.jetbrains.annotations.Nullable;
+
+public class EvtPreSpawn extends SkriptEvent {
+
+    private static final boolean HAS_PRE_CREATURE_SPAWN_EVENT = Skript.classExists("com.destroystokyo.paper.event.entity.PreCreatureSpawnEvent");
+    private static final boolean HAS_PRE_SPAWNER_SPAWN_EVENT = Skript.classExists("com.destroystokyo.paper.event.entity.PreSpawnerSpawnEvent");
+    private static final boolean HAS_PRE_PHANTOM_SPAWN_EVENT = Skript.classExists("com.destroystokyo.paper.event.entity.PhantomPreSpawnEvent");
+
+    static {
+
```
Comment (ShaneBeee) [inline src/main/java/com/shanebeestudios/skbee/elements/other/events/EvtPreSpawn.java:29]:
```suggestion
```

File: src/main/java/com/shanebeestudios/skbee/elements/other/events/EvtPreSpawn.java [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+        // Paper - PreCreatureSpawnEvent
+        if (HAS_PRE_CREATURE_SPAWN_EVENT) {
+
```
Comment (ShaneBeee) [inline src/main/java/com/shanebeestudios/skbee/elements/other/events/EvtPreSpawn.java:32]:
```suggestion
```

File: src/main/java/com/shanebeestudios/skbee/elements/other/events/EvtPreSpawn.java [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+            Skript.registerEvent("Pre Creature Spawn", EvtPreSpawn.class, PreCreatureSpawnEvent.class,
+                            "pre [creature] spawn[ing] [of %entitydatas%]")
+                    .description("Called before an entity is spawned into the world.",
+                            "Note: The spawning entity does not exist when this event is called only the entitytype exists.",
+                            "",
+                            "event-spawnreason - the reason the entity is spawned",
+                            "event-location - the location the spawned entity will appear",
+                            "event-entitytype - the type of entity being spawned")
```
Comment (ShaneBeee) [inline src/main/java/com/shanebeestudios/skbee/elements/other/events/EvtPreSpawn.java:40]:
```suggestion
                            "\nevent-spawnreason = the reason the entity is spawned.",
                            "\nevent-location = the location the spawned entity will appear.",
                            "\nevent-entitytype = the type of entity being spawned.")
```

File: src/main/java/com/shanebeestudios/skbee/elements/other/events/EvtPreSpawn.java [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+                            "\tbroadcast \"a %event-entitytype% is spawning in\"")
+                    .since("INSERT VERSION");
+
+            EventValues.registerEventValue(PreCreatureSpawnEvent.class, Location.class, new Getter<>() {
+                @Override
+                public Location get(PreCreatureSpawnEvent event) {
+                    return event.getSpawnLocation();
+                }
+            }, EventValues.TIME_NOW);
+            EventValues.registerEventValue(PreCreatureSpawnEvent.class, EntityData.class, new Getter<>() {
```
Comment (ShaneBeee) [inline src/main/java/com/shanebeestudios/skbee/elements/other/events/EvtPreSpawn.java:51]:
```suggestion
            }, EventValues.TIME_NOW);

            EventValues.registerEventValue(PreCreatureSpawnEvent.class, EntityData.class, new Getter<>() {
```

**Comment by Fusezion** [comment]
Honestly the pre spawn event is called almost as much as block physics so I will need to add a warning about that to it

File: src/main/java/com/shanebeestudios/skbee/elements/other/events/EvtPreSpawn.java [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+                @Override
+                public SpawnReason get(PreCreatureSpawnEvent event) {
+                    return event.getReason();
+                }
+            }, EventValues.TIME_NOW);
+        }
+
+        // Paper - PreSpawnerSpawnEvent
+        if (HAS_PRE_SPAWNER_SPAWN_EVENT) {
+
```
Comment (ShaneBeee) [inline src/main/java/com/shanebeestudios/skbee/elements/other/events/EvtPreSpawn.java:68]:
```suggestion
```

File: src/main/java/com/shanebeestudios/skbee/elements/other/events/EvtPreSpawn.java [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+            Skript.registerEvent("Pre Spawner Spawn", EvtPreSpawn.class, PreSpawnerSpawnEvent.class,
+                            "pre spawner spawn[ing] [of %entitydatas%]")
+                    .description("Called before an entity is spawned via a spawner.",
+                            "Note: The spawned entity does not exist when this event is called only the entitytype exists.",
+                            "",
+                            "View the pre creature spawn event for more event values.",
+                            "event-block - the block location of the spawner spawning the entity")
```
Comment (ShaneBeee) [inline src/main/java/com/shanebeestudios/skbee/elements/other/events/EvtPreSpawn.java:75]:
```suggestion
                            "\nView the pre creature spawn event for more event values.",
                            "\nevent-block = the block location of the spawner spawning the entity.")
```

File: src/main/java/com/shanebeestudios/skbee/elements/other/events/EvtPreSpawn.java [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+
+        // Paper - PhantomPreSpawnEvent
+        if (HAS_PRE_PHANTOM_SPAWN_EVENT) {
+            Skript.registerEvent("Pre Phantom Spawn", EvtPreSpawn.class, PhantomPreSpawnEvent.class,
+                    "pre phantom spawn[ing]")
+                    .description("Called before a phantom is spawned for an entity.",
+                            "Note: The phantom entity does not exist when this event is called only the entitytype exists.",
+                            "",
+                            "View the pre creature spawn event for more event values.",
+                            "event-entity - the entity the spawned phantom is spawning for")
```
Comment (ShaneBeee) [inline src/main/java/com/shanebeestudios/skbee/elements/other/events/EvtPreSpawn.java:96]:
```suggestion
                            "\nView the pre creature spawn event for more event values.",
                            "\nevent-entity = the entity the spawned phantom is spawning for.")
```

**Review by ShaneBeee — CHANGES_REQUESTED** [review]
just a few tiny little things

**Review by ShaneBeee — APPROVED** [review]

**Pull request merged**